src/web/js/output-ui.js

```diff
@@ -134,278 +218,55 @@
-          if (warnDesired === 0) {
-            obj[0].style.pointerEvents = "none";
-          } else {
-            obj[0].style.pointerEvents = "all";
-          }
-          if (warnDesired === fadeAmt) {
-            obj.fadeTo("fast", fadeAmt, function() {
-              setTimeout(function() {
-                obj.fadeTo("slow", 0.0);
-                warnDesired = 0;
-              }, warnDuration);});
-          } else {
-            obj.fadeTo("fast", 0.0);
-          }
-        }
-      }
-    }
-
-    function hintLoc(runtime, editors, srcloc, loc) {
-      var editor = editors[runtime.getField(loc, "source")];
-      if(!editor) { return; }
-      var cmLoc = cmPosFromSrcloc(runtime, srcloc, loc);
-      var locKey = cmlocToCSSClass(cmLoc);
-      var view = editor.getScrollInfo();
-      var charCh = editor.charCoords(cmLoc.start, "local");
-      $("."+locKey).addClass("hover");
-      if (view.top > charCh.top) {
-        warnDesired = fadeAmt;
-        var warningUpper = jQuery(editor.getWrapperElement()).find(".warning-upper");
-        setTimeout(function() {setWarningState(warningUpper);}, warnWait);
-      } else if (view.top + view.clientHeight < charCh.bottom) {
-        warnDesired = fadeAmt;
-        var warningLower = jQuery(editor.getWrapperElement()).find(".warning-lower");
-        setTimeout(function() {setWarningState(warningLower);}, warnWait);
-      }
-    }
-
-    function unhintLoc(runtime, editors, srcloc, loc) {
-      var editor = editors[runtime.getField(loc, "source")];
-      if(!editor) { return; }
-      var cmLoc = cmPosFromSrcloc(runtime, srcloc, loc);
-      var locKey = cmlocToCSSClass(cmLoc);
-      $("."+locKey).removeClass("hover");
-      warnDesired = 0;
-      setTimeout(function() { setWarningState(jQuery(".warning-upper"));},
-                 warnWait);
-      setTimeout(function() { setWarningState(jQuery(".warning-lower"));},
-                 warnWait);
-    }
-
-    function gotoLoc(runtime, editors, srcloc, loc) {
-      var editor = editors[runtime.getField(loc, "source")];
-      if(!editor) { return; }
-      var cmLoc = cmPosFromSrcloc(runtime, srcloc, loc);
-      warnDesired = 0;
-      jQuery(".warning-upper").fadeOut("fast");
-      jQuery(".warning-lower").fadeOut("fast");
-      editor.scrollIntoView(cmLoc.start, 100)
-    }
-
-    function hoverSrclocAnchor(runtime, editors, srcloc, elt, loc) {
-      var warnDesired = 0;
-      var warnWait = 250;
-      var warnDuration = 5000;
-      var fadeAmt = 0.5;
+    function hintLoc(position) {
+      $(".warning-upper.hinting, .warning-lower.hinting").removeClass("hinting");
+      
+      var editor = position.doc.getEditor();
+      
+      if (!(editor instanceof CodeMirror))
+        throw new Error("Source location not in editor", position);
+      
+      var coord = editor.charCoords(
+        {line: position.from.line, ch:0},
+        position.source === "definitions://" ? "local" : "page");
```

Comment:
It would be nice to make these ad hoc enums be constants defined elsewhere, as e.g. `LOCALSOURCE = "local", PAGESOURCE = "page"`, since presumably they are duplicated string literals.